broadcast/src/buffered/engine.rs

```diff
@@ -261,13 +326,37 @@ impl<
     /// Returns `true` if the message was inserted, `false` if it was already present.
     /// Updates the deque, item count, and message cache, potentially evicting an old message.
     fn insert_message(&mut self, peer: P, msg: M) -> bool {
+        // Get the identity and digest of the message
+        let identity = msg.identity();
         let digest = msg.digest();
 
         // Send the message to the waiters, if any, ignoring errors (as the receiver may have dropped)
-        if let Some(responders) = self.waiters.remove(&digest) {
-            for responder in responders {
+        if let Some(mut waiters) = self.waiters.remove(&identity) {
+            let mut i = 0;
+            while i < waiters.len() {
+                // Get the sender and digest filters
+                let (sender_filter, digest_filter, _) = &waiters[i];
+
+                // Keep the waiter if either filter does not match.
+                if sender_filter.as_ref().is_some_and(|s| s != &peer)
+                    || digest_filter.is_some_and(|d| d != digest)
+                {
+                    i += 1;
+                    continue;
+                }
+
+                // Filters match → fulfil the subscription and drop the entry.
+                //
+                // The index `i` is intentionally not incremented here to check
+                // the element that was swapped into position `i`.
+                let (_, _, responder) = waiters.swap_remove(i);
                 self.respond_subscribe(responder, msg.clone());
             }
```

Comment:
Style: prefer using [`retain_mut`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.retain_mut) instead of manually iterating and moving items around.

Reply:
See note below about it not working.